File: javascript/simple-cipher/simple-cipher.js
export class Cipher {
  constructor(key = null) {
    if (key === null) {
      for (let i = 0; i <= 100; i++) {
        key += String.fromCharCode(Math.floor(Math.random() * 26) + 'a'.charCodeAt(0)); 
      }
    } else {
      if (key === '') throw new Error('Bad key');
      for (let i = 0; i < key.length; i++) {
        if (!key[i].match(/^[a-z]+$/)) {
          throw new Error('Bad key');
        }
      }
    }
    this.key = key;
  }

  applyShift(char, index, isEncoding) {
    const min = 'a'.charCodeAt(0);
    const max = 'z'.charCodeAt(0);
    const keyCode = this.key.charCodeAt(index % this.key.length);
    const shift = keyCode - min;
    const charCode = char.charCodeAt(0);
    let newCode;
    if (isEncoding) {
      newCode = charCode + shift;
      if (newCode > max) {
        newCode = min + (newCode - max) - 1;
      }
    } else {
      newCode = charCode - shift;
      if (newCode < min) {
        newCode = max - (min - newCode) + 1;
      }
    }
    return String.fromCharCode(newCode)
  }

  encode(s) {
    let encodedString = '';
    for (let i = 0; i < s.length; i++) {
      encodedString += this.applyShift(s[i].toLowerCase(), i, true);
    }
    return encodedString;
  }

  decode(s) {
    let decodedString = '';
    for (let i = 0; i < s.length; i++) {
      decodedString += this.applyShift(s[i].toLowerCase(), i, false);
    }
    return decodedString;
  }
}

File: python/simple-cipher/simple_cipher.py
from random import randint

class Cipher(object):
    def __init__(self, key=None):
        if key == None:        
            key = ''.join(chr(randint(ord('a'), ord('z'))) for _ in range(100))
        self.key = key

    def encode(self, text):
        arr = list(text)
        for i, char in enumerate(arr):
            shift = ord(self.key[i%len(self.key)]) - ord('a')
            arr[i] = chr(((ord(char) - ord('a') + shift) % 26) + ord('a'))
        return ''.join(arr)
        # one-liner:
        # return ''.join(chr(((ord(char) - ord('a') + ord(self.key[i%len(self.key)]) - ord('a')) % 26) + ord('a')) for i, char in enumerate(list(text)))

    def decode(self, text):
        arr = list(text)
        for i, char in enumerate(arr):
            shift = ord(self.key[i%len(self.key)]) - ord('a')
            arr[i] = chr(((ord(char) - shift - ord('a') + 26) % 26) + ord('a'))
        # one-liner:
        # return ''.join(chr(((ord(char) - (ord(self.key[i%len(self.key)]) - ord('a')) - ord('a') + 26) % 26) + ord('a')) for i, char in enumerate(list(text)))